core/src/io.rs

```diff
@@ -475,9 +480,33 @@ fn read_request(request_str: &str) -> Result<Request, Error> {
 	crate::serde_from_str(request_str).map_err(|_| Error::new(ErrorCode::ParseError))
 }
 
-fn write_response(response: Response) -> String {
-	// this should never fail
-	serde_json::to_string(&response).unwrap()
+fn write_response(response: WrapResponse) -> String {
+	match response {
+		WrapResponse::Single(output) => output.response,
+		WrapResponse::Batch(outputs) => {
+			if outputs.len() == 0 {
+				return String::from("[]");
+			}
+
+			// "[" and comma-separated outputs and "]"
+			let commas = outputs.len() - 1;
+			let brackets = 2;
+			let length = brackets + commas + outputs.iter().map(|output| output.response.len()).sum::<usize>();
+
+			// Not elegant, but guaranteed to only copy responses once
+			let mut response = String::with_capacity(length);
+			response.push('[');
+			for output in outputs {
+				if response.len() > 1 {
+					response.push(',');
+				}
+				response.push_str(&output.response);
+			}
+			response.push(']');
```

Comment:
Since you guarantee that there is always at least one element in the array, you can skip branching if you always push a comma at the end, and replace the last one with a bracket:

```suggestion
			for output in outputs {
				response.push_str(&output.response);
				response.push(',');
			}
			response.pop();
			response.push(']');
```